keywrap/keyprovider/keyprovider.go

```diff
@@ -0,0 +1,260 @@
+				if err != nil {
+					return nil, errors.Wrap(err, "error while retrieving keyprovider protocol grpc output")
+				}
+
+				return protocolOuput.KeyWrapResults.Annotation, nil
+			} else {
+				return nil, errors.New("unsupported keyprovider invocation. supported invocation methods are grpc and cmd")
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+// UnwrapKey wraps reads out the OCICRYPT_KEYPROVIDER_CONFIG env variable and calls appropriate binary executable/grpc server for unwrapping the session key based on the protocol given in annotation for recipients and gets decrypted optsData,
+// which describe the symmetric key used for decrypting the layer
+func (kw *keyProviderKeyWrapper) UnwrapKey(dc *config.DecryptConfig, jsonString []byte) ([]byte, error) {
+
+	input, err := json.Marshal(KeyProviderKeyWrapProtocolInput{
+		Operation: OpKeyWrap,
+		KeyUnwrapParams: KeyUnwrapParams{
+			Dc:         dc,
+			Annotation: jsonString,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	providersMap := kw.args.(map[string]interface{})
+	if _, ok := providersMap["cmd"]; ok {
+		protocolOuput, err := getProviderCommandOutput(input, providersMap)
+		if err != nil {
+			// If err is not nil, then ignore it and continue with rest of the given keyproviders
+			return nil, err
+		}
+
+		return protocolOuput.KeyUnwrapResults.OptsData, nil
+	} else if socketFile, ok := providersMap["grpc"]; ok {
+		protocolOuput, err := getProviderGRPCOutput(input, socketFile, OpKeyUnwrap)
+		if err != nil {
+			// If err is not nil, then ignore it and continue with rest of the given keyproviders
+			return nil, err
+		}
+
+		return protocolOuput.KeyUnwrapResults.OptsData, nil
+	} else {
+		return nil, errors.New("unsupported keyprovider invocation. supported invocation methods are grpc and cmd ")
+	}
+}
+
+func getProviderGRPCOutput(input []byte, socketFile interface{}, operation KeyProviderKeyWrapProtocolOperation) (*KeyProviderKeyWrapProtocolOuput, error) {
+	socketFileStr := socketFile.(string)
+	var protocolOuput KeyProviderKeyWrapProtocolOuput
+	var grpcOutput *keyproviderpb.KeyProviderKeyWrapProtocolOutput
+	cc, err := grpc.Dial(socketFileStr, grpc.WithInsecure())
+	defer func() {
+		derr := cc.Close()
+		if derr != nil {
+			log.WithError(derr).Error("Error closing grpc socket")
+		}
+	}()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "error while dialing rpc server")
+	}
+
+	client := keyproviderpb.NewKeyProviderServiceClient(cc)
+	req := &keyproviderpb.KeyProviderKeyWrapProtocolInput{
+		KeyProviderKeyWrapProtocolInput: input,
+	}
+
+	if operation == OpKeyWrap {
+		grpcOutput, err = client.WrapKey(context.Background(), req)
+		if err != nil {
+			return nil, errors.Wrap(err, "Error from grpc method")
+		}
+	} else if operation == OpKeyUnwrap {
+		grpcOutput, err = client.UnWrapKey(context.Background(), req)
+		if err != nil {
+			return nil, errors.Wrap(err, "Error from grpc method")
+		}
+	}
```

Comment:
do we need to check if its not either case to return err? `switch` statement would work as well if you prefer

Reply:
no I think. I dont expect a scenerio/possibility as such